### PR TITLE
Prevent duplicates in module dependency lists

### PR DESCRIPTION
**Pull request by msullivan**
Currently in some cases we will write out a lot of duplicate entries
in the dependency lists of modules. Refactor dependency management to
prevent this.

### CODE REVIEW

File: mypy/build.py [hunk 1]
```diff
@@ -2157,11 +2182,11 @@ def _patch_indirect_dependencies(self,
...
                 continue
-            if dep not in self.suppressed and dep not in self.manager.missing_modules:
-                self.dependencies.append(dep)
+            if dep not in self.suppressed_set and dep not in self.manager.missing_modules:
+                self.add_dependency(dep)
                 self.priorities[dep] = PRI_INDIRECT
-            elif dep not in self.suppressed and dep in self.manager.missing_modules:
-                self.suppressed.append(dep)
+            elif dep in self.manager.missing_modules:
+                self.suppress_dependency(dep)
```
Comment (ilevkivskyi) [inline mypy/build.py]: TBH it is a bit hard to follow logic here. Why is it OK to remove the `dep not in self.suppressed` from the check for this branch?

Comment (msullivan) [reply mypy/build.py]: Because the check is pushed down into `suppress_dependency`; I can add it back if it seems clearer, though

Comment (ilevkivskyi) [reply mypy/build.py]: I would keep it (just in case if anything is changed that method). Also I noticed you keep other similar checks, so it will add some consistency (or are these are checks actually needed, for example the one on line 2777, green side).
